<% brew::brew(get.path2template("page.header")) %>
<%
source(paste(IRAP.DIR,"aux/R","irap_misc.R",sep="/"))
cur.plot<-0
methods.colors <- rainbow(nfiles)
names(methods.colors)<-colnames(data2html$p.cor)

ERR.LABEL <- "Error (%)"
errors<-diag(data2html$quant.ndiff)
is.Simulated <- F
if ( sum(errors[!is.na(errors)]) > 0 ) {
  # sort the genes by average error accross all methods
  germ<-rowMeans(data2html$gene.nerr)
  o<-sort(germ,decreasing=FALSE,index.return=T)
  data2html$gene.nerr.sort <- data2html$gene.nerr[o$ix,]
  germ <- germ[o$ix]
  is.Simulated <- T
}
%>
<%# Comment -- wrap up everything in a div %>
<!--<DIV name='de_res'>-->
<H1><%=title%></H1>
<H2><%=sources.menu%></H2>
<H3 class="section">Overall Comparison <%=opt$feature%> quantification</H3>


<!-- quant p.cor + rank  rank.diff -->
<table class="center">
<tr>
<TD>
<%
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_counts_pairs.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Pearson correlation of gene expression (counts) estimates (red line corresponds to the lowess curve generated).",
      to.plot=function() {
#log="xy" broken
            pairs(data2html$gene.expr, lower.panel=panel.smooth, upper.panel=panel.cor,pch=".")
        })
cat(r.a$html)
%>
<%=cur.plot%>
</TD>
</tr>

<tr>
<TD>
<%
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_rpkms_pairs.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Pearson correlation of gene expression (RPKMs) estimates  (red line corresponds to the lowess curve generated).",
          to.plot=function() {
            pairs(data2html$gene.rpkms, lower.panel=panel.smooth, upper.panel=panel.cor,pch=".")
        })
cat(r.a$html)
%>
<%=cur.plot%>
</TD>
</tr>

<tr>
<TD>
<%
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_sp_counts_pairs.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Spearman correlation of gene expression (counts) estimates (red line corresponds to the lowess curve generated).",
      to.plot=function() {
#log="xy" broken
            pairs(data2html$gene.expr, lower.panel=panel.smooth, upper.panel=panel.cor.spearman,pch=".")
        })
cat(r.a$html)
%>
<%=cur.plot%>
</TD>
</tr>

<tr>
<TD>
<%
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_sp_rpkms_pairs.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Spearman correlation of gene expression (RPKMs) estimates  (red line corresponds to the lowess curve generated).",
          to.plot=function() {
            pairs(data2html$gene.rpkms, lower.panel=panel.smooth, upper.panel=panel.cor.spearman,pch=".")
        })
cat(r.a$html)
%>
<%=cur.plot%>
</TD>
</tr>

<TR><TD colspan=2>
<H4 class="section"><H4 class="section">Average Spearman's correlation</H4>  
</TD></TR>

<tr><td>
<%

cur.plot<-cur.plot+1
#save.image(paste(file.prefix,".Rdata",sep=""))
#sort the columns and rows
#pinfo("Generating ",file.prefix,"_quant_sp_corr2.png")
r.a<-gen.plot2report(filename=paste(file.prefix,"_quant_sp_corr2.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$p.cor,ncolors=(1+1)*10,dendrogram="none",main=paste("Raw counts - Spearman correlation",sep=""),density.info="none",Rowv=FALSE,Colv=FALSE,cexRow=0.65,cexCol=0.65)
      #ColSideColors=methods.colors[rownames(data2html$p.cor)])
        })
cat(r.a$html)
%>
<%=cur.plot%>
</TD>
</tr>


<tr>
<TD>
<%
#pinfo("Generating ",file.prefix,"_rank_sp_corr2.png")
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_rank_sp_corr2.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$rank.cor,ncolors=(1+1)*10,main=paste("Rank - Spearman correlation",sep=""),density.info="none",dendrogram="none",Rowv=FALSE,Colv=FALSE,cexRow=0.65,cexCol=0.65)
        })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD></TR>

<tr><td>
<%
cur.plot<-cur.plot+1
#pinfo("Generating ",file.prefix,"_quant_sp_corr.png")
r.a<-gen.plot2report(filename=paste(file.prefix,"_quant_sp_corr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$p.cor,ncolors=(1+1)*10,main=paste("Raw counts - Spearman correlation",sep=""),density.info="none",cexRow=0.65,cexCol=0.65)
        })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD>
</tr><tr>
<TD>
<%
cur.plot<-cur.plot+1
#pinfo("Generating ",file.prefix,"_rank_sp_corr.png")
r.a<-gen.plot2report(filename=paste(file.prefix,"_rank_sp_corr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$rank.cor,ncolors=(1+1)*10,main=paste("Rank - Spearman correlation",sep=""),density.info="none",cexRow=0.65,cexCol=0.65)
        })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD></TR>

<tr>
<TD>
<%
cur.plot<-cur.plot+1
ncomp<-5
ngenes<-8
sel.genes<-selected.examples(ncomp,ngenes)
for (c in names(sel.genes)) {
  cs <- strsplit(c,pipeline.sep)
  comp1 <- cs[[1]][1]
  comp2 <- cs[[1]][2]
#  dev.new(width=15,height=5)
  r.a<-gen.plot2report(filename=paste(file.prefix,"_",c,"_selection.png",sep=""),
          width=900,
          height=300,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              plot.selection(comp1,comp2,sel.genes[[c]],data2html$gene.expr,base.mean.deseq[[c]],log2.fold.change.deseq[[c]],p.val.deseq[[c]],p.val.ttest[[c]],fdr)
        })
  cat(r.a$html)
 }

%>
P<%=cur.plot%>
</TD></TR>

<TR><TD colspan=2>
<H4 class="section">Quantification normalized difference/error</H4>  
</TD></TR>

<TR><TD>
<%
cur.plot<-cur.plot+1
#pinfo("Generating ",file.prefix,"_ndiff_sp_corr.png")
r.a<-gen.plot2report(filename=paste(file.prefix,"_ndiff_sp_corr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$quant.ndiff,ncolors=20,main=paste("Mean difference (%)",sep=""),density.info="none",cexRow=0.65,cexCol=0.65,ColSideColors=methods.colors[rownames(data2html$p.cor)])
        })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD></tr>

<tr><TD>
<%
#Error
cur.plot<-cur.plot+1
ERR.LABEL <- "Error (%)"
errors<-diag(data2html$quant.ndiff)
if ( is.Simulated ) {
  # sort the genes by average error accross all methods
  germ<-rowMeans(data2html$gene.nerr)
  o<-sort(germ,decreasing=FALSE,index.return=T)
  data2html$gene.nerr.sort <- data2html$gene.nerr[o$ix,]
  germ <- germ[o$ix]
#  width=max((length(errors)*70),400),
#  height=500,
  r.a<-gen.plot2report(filename=paste(file.prefix,"_ndiff_error_bp.png",sep=""),
  height=max((length(errors)*60),400),
  width=500,
  html=TRUE,
  ps=TRUE,
  caption="Normalized error",
  to.plot=function() {
    par(mar=c(5,10,4,4))
    boxplot(data2html$gene.nerr.sort,nonotch=T,outline=F,ylim=c(0,100),las=2,xlab=ERR.LABEL,col=methods.colors[colnames(data2html$gene.nerr.sort)],horizontal=T)
  })
  par(mar=c(5,5,4,4))
  cat(r.a$html)
} else {
   pinfo("Skipping generation of normalized error plot - no data")
}

%>
P<%=cur.plot%>
</TD></TR>
<tr><TD>
<%
cur.plot<-cur.plot+1
# Include the outliers
if ( is.Simulated ) {
  r.a<-gen.plot2report(filename=paste(file.prefix,"_ndiff_error_bpo.png",sep=""),
  height=max((length(errors)*60),400),
  width=500,
  html=TRUE,
  ps=TRUE,
  caption="Normalized error",
  to.plot=function() {
    par(mar=c(5,10,4,4))
    boxplot(data2html$gene.nerr.sort,nonotch=T,outline=T,las=2,xlab=ERR.LABEL,col=methods.colors[colnames(data2html$gene.nerr.sort)],horizontal=T)
  })
  par(mar=c(5,5,4,4))
  cat(r.a$html)
} 
%>
P<%=cur.plot%>
</TD></TR>
<tr>
<TD>
<%
cur.plot<-cur.plot+1
#pinfo("Generating ",file.prefix,"_gene_methods_nerrByexpr.png")
if ( is.Simulated) {
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_nerrByexpr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
          caption="Genes sorted by average normalized error (accross all pipelines)",
	  ps=TRUE,
          to.plot=function() {            
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
#           plot(smooth.spline(data2html$gene.nerr.sort[,1]),type="l",col=methods.colors[1],ylab=ERR.LABEL,xlab="Genes",log="y",ylim=c(0,200),xaxt="n")
            plot(smooth.spline(data2html$gene.nerr.sort[,1]),type="l",col=methods.colors[1],ylab=ERR.LABEL,xlab="Genes",ylim=c(0,200),xaxt="n")
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
                   lines(smooth.spline(data2html$gene.nerr.sort[,i]),col=methods.colors[colnames(data2html$gene.nerr.sort)[i]])
            }
            lines(smooth.spline(germ),type="l",col="black")
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=append(colnames(data2html$gene.nerr.sort),"Mean"), lty=1, col=append(methods.colors[colnames(data2html$gene.nerr.sort)],"black"), title="", bty='n')
            axis(1,at=seq(1,nrow(data2html$gene.nerr.sort)+1,nrow(data2html$gene.nerr.sort)/5),labels=paste(seq(0,100,100/5),"%",sep=""))
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
if ( is.Simulated) {
cur.plot<-cur.plot+1
sq<-names(sort(data2html$true.quant.mean.f))
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_nerrByexpr2.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Genes sorted by true quantification",
          to.plot=function() {#,
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
            plot(smooth.spline(data2html$gene.nerr.sort[sq,1]),type="l",ylab=ERR.LABEL,xlab="Genes",ylim=c(0,200),col=methods.colors[1],xaxt="n")
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
                   lines(smooth.spline(data2html$gene.nerr.sort[sq,i]),col=methods.colors[colnames(data2html$gene.nerr.sort)[i]])
            }
            #lines(data2html$true.quant.mean.f[sq]/sum(data2html$true.quant.mean.f)*length(data2html$true.quant.mean.f)*100,col="gray")
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
            axis(1,at=seq(1,nrow(data2html$gene.nerr.sort)+1,nrow(data2html$gene.nerr.sort)/5),labels=paste(seq(0,100,100/5),"%",sep=""))

      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
if ( is.Simulated ) {
cur.plot<-cur.plot+1
true.rpkms<-counts2RPKMs(as.matrix(data2html$true.quant.mean.f),annot.table)
names(true.rpkms)<-names(data2html$true.quant.mean.f)
sq<-names(sort(true.rpkms))
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_nerrByexpr2rpkm.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Genes sorted by true quantification (RPKM)",
          to.plot=function() {#,
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
            plot(y=data2html$gene.nerr.sort[sq,1],x=true.rpkms[rownames(data2html$gene.nerr.sort[sq,])],type="p",ylab=ERR.LABEL,xlab="RPKMs",ylim=c(0,300),col=methods.colors[1],pch=".",xlim=c(0,max(true.rpkms)))
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
                   points(y=data2html$gene.nerr.sort[sq,i],x=true.rpkms[rownames(data2html$gene.nerr.sort[sq,])],col=methods.colors[i])
            }
            #lines(data2html$true.quant.mean.f[sq]/sum(data2html$true.quant.mean.f)*length(data2html$true.quant.mean.f)*100,col="gray")
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
## if ( sum(errors) > 0 ) {
## cur.plot<-cur.plot+1
## sq<-as.character(annot.table$ID[as.character(annot.table$ID[sort(annot.table$len,index.return=T)$ix]) %in% rownames(data2html$gene.nerr)])

## r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_len_nerrByexpr.png",sep=""),
##           width=500,
##           height=500,
## 	  html=TRUE,
## 	  ps=TRUE,
##           caption="Genes sorted by length",
##           to.plot=function() {#,
## 	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
##             plot(smooth.spline(data2html$gene.nerr[sq,1]),type="l",ylab=ERR.LABEL,xlab="Genes",ylim=c(0,200),col=methods.colors[1],xaxt="n")
##             for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
##                    lines(smooth.spline(data2html$gene.nerr.sort[sq,i]),col=methods.colors[colnames(data2html$gene.nerr.sort)[i]])
##             }
##             lines(data2html$true.quant.mean.f[sq]/sum(data2html$true.quant.mean.f)*length(data2html$true.quant.mean.f)*100,col="gray")
## 	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
##             axis(1,at=seq(1,nrow(data2html$gene.nerr.sort)+1,nrow(data2html$gene.nerr.sort)/5),labels=paste(seq(0,100,100/5),"%",sep=""))

##       })
## cat(r.a$html)
## }
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
if ( is.Simulated ) {
cur.plot<-cur.plot+1
sq<-names(sort(data2html$true.quant.mean.f))
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_nerrByexpr3.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Error vs true quantification",
          to.plot=function() {#,
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
            plot(smooth.spline(data2html$gene.nerr.sort[sq,1],data2html$true.quant.mean.f[sq]),type="l",ylab=ERR.LABEL,xlab="Counts (true)",ylim=c(0,200),col=methods.colors[1],log="x")
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
                   lines(smooth.spline(data2html$gene.nerr.sort[sq,i],data2html$true.quant.mean.f[sq]),col=methods.colors[colnames(data2html$gene.nerr.sort)[i]])
            }
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
     
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>


<tr>
<TD>
<%
#pinfo("Generating ",file.prefix,"_expr_acc_nerr.png")
if ( is.Simulated ) {  
cur.plot<-cur.plot+1
  #save.image("/homes/nf/lixo.Rdata")
  quant.vals.u<-sort(unique(round(data2html$true.quant.mean.f,0)))
  quant.vals.r<-round(data2html$true.quant.mean.f,0)
  r.a<-gen.plot2report(filename=paste(file.prefix,"_expr_acc_nerr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              my.filt<-function(val,l1,l2) {
                sum(l2[l1>=val])
              }
 	    terror<-sum(data2html$gene.nerr.sort[,1])
            err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,1])
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(smooth.spline(err.by.expr/terror*100),type="l",log="x",xlab="Quantification (reads)",ylab="Accumulative error ( X >= error )",ylim=c(0,100),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
 	      terror<-sum(data2html$gene.nerr.sort[,i])
              err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,i])
              lines(smooth.spline(err.by.expr/terror*100),col=methods.colors[i])
            }
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
#pinfo("Generating ",file.prefix,"_expr_acc_nerr.png")
if ( is.Simulated ) {  
cur.plot<-cur.plot+1
  #save.image("/homes/nf/lixo.Rdata")
  quant.vals.u<-sort(unique(round(data2html$true.quant.mean.f,0)))
  quant.vals.r<-round(data2html$true.quant.mean.f,0)
  r.a<-gen.plot2report(filename=paste(file.prefix,"_expr_acc_nerr2.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              my.filt<-function(val,l1,l2) {
                sum(l2[l1<=val])
              }
 	    terror<-sum(data2html$gene.nerr.sort[,1])
            err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,1])
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(smooth.spline(err.by.expr/terror*100),type="l",log="x",xlab="Quantification (reads)",ylab="Accumulative error ( X <= error )",ylim=c(0,100),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
 	      terror<-sum(data2html$gene.nerr.sort[,i])
              err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,i])
              lines(smooth.spline(err.by.expr/terror*100),col=methods.colors[i])
            }
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
if ( is.Simulated ) {  
  cur.plot<-cur.plot+1
  quant.vals.u<-sort(unique(round(data2html$true.quant.mean.f,0)))
  quant.vals.r<-round(data2html$true.quant.mean.f,0)
  r.a<-gen.plot2report(filename=paste(file.prefix,"_expr_nerr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
	  caption="Average error VS expression",
          to.plot=function() {
              my.filt<-function(val,l1,l2) {
                mean(l2[l1==val])
              }
            err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,1])
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(smooth.spline(err.by.expr),type="l",log="x",xlab="Quantification (reads)",ylab="avg. error",ylim=c(0,100),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort)))) {
              err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.nerr.sort[,i])
              lines(smooth.spline(err.by.expr),col=methods.colors[colnames(data2html$gene.nerr.sort)[i]])
            }
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>
<tr>
<TD>
<%
if ( is.Simulated ) {  
  cur.plot<-cur.plot+1
  r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_nerrVsexpr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(data2html$true.quant.mean.f,data2html$gene.nerr.sort[,1],log="xy",type="p",pch=19,xlab="True gene quantification (reads)",ylab=ERR.LABEL,col=methods.colors[1],cex=0.3)
            for ( i in seq(2,length(colnames(data2html$gene.nerr.sort))) ) { points(data2html$true.quant.mean.f,data2html$gene.nerr.sort[,i],col=methods.colors[i],pch=19,cex=0.3) }
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), pch=19, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%
nerrors<-diag(data2html$quant.ndiff)
nttest<-diag(data2html$ttest.ndiff)
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_ttestvnserr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Number of significantly different values using paired-ttest adjusted using Benjamini and Hochberg
procedure with a fdr=0.01",
          to.plot=function() {
	      par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
              plot(x=nerrors,y=nttest,type="p",pch=19,xlab="Error",ylab="Sig. different",main=paste("","ttest (fdr=0.01)",sep=""),xlim=c(0,max(nerrors)),col=methods.colors[names(nerrors)])
              # Add legend to top right, outside plot region
	      legend("topright", inset=c(-0.6,0), legend=names(nerrors), pch=19, col=methods.colors[names(nerrors)], title="", bty='n')
      })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD>

<tr>
<TD>
<%
if ( is.Simulated ) {
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_deseqvnserr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Number of significantly different values using DEseq (fdr 0.01)",
          to.plot=function() {
	      par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend

              plot(x=data2html$quant.ndiff,y=data2html$sig.diff,type="p",pch=19,xlab="Error",ylab="Sig. different",main=paste("","DESeq (fdr=0.01)",sep=""),col=methods.colors[names(nerrors)],xlim=c(0,max(data2html$quant.ndiff)))
              # Add legend to top right, outside plot region
	      legend("topright", inset=c(-0.6,0), legend=names(nerrors), pch=19, col=methods.colors[names(nerrors)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</TD>

<!-- raw difference -->
<TR><TD colspan=2>
<H4 class="section">Quantification difference/error</H4>  
</TD></TR>

<TR><TD>
<%
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_diff_sp_corr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              gen.heatmap(data2html$quant.diff,ncolors=20,main=paste("Raw counts - Mean difference (%)",sep=""),density.info="none",cexRow=0.65,cexCol=0.65)
        })
cat(r.a$html)
%>
P<%=cur.plot%>
</TD></tr><tr><TD>
<%
#Error
ERR.LABEL <- "Error (%)"
errors<-diag(data2html$quant.diff)
if ( is.Simulated ) {
  cur.plot<-cur.plot+1
  # sort the genes by average error accross all methods
  germ<-rowMeans(data2html$gene.err)
  o<-sort(germ,decreasing=FALSE,index.return=T)
  data2html$gene.err.sort <- data2html$gene.err[o$ix,]
  germ <- germ[o$ix]
  r.a<-gen.plot2report(filename=paste(file.prefix,"_diff_error_bp.png",sep=""),
  width=max((length(errors)*70),400),
  height=500,
  html=TRUE,
  ps=TRUE,
  to.plot=function() {
    par(mar=c(5,10,4,4))
    boxplot(data2html$gene.err.sort,nonotch=T,outline=F,las=2,xlab=ERR.LABEL,col=methods.colors[colnames(data2html$gene.err.sort)],horizontal=T)
  })
  par(mar=c(5,5,4,4))
  cat(r.a$html)
}
%>
P<%=cur.plot%>
</TD></TR>
<tr><TD>
<%
if ( is.Simulated ) {
  cur.plot<-cur.plot+1
  r.a<-gen.plot2report(filename=paste(file.prefix,"_diff_error_bpo.png",sep=""),
  height=max((length(errors)*70),400),
  width=500,
  html=TRUE,
  ps=TRUE,
  to.plot=function() {
    par(mar=c(5,10,4,4))
    boxplot(data2html$gene.err.sort,nonotch=T,outline=T,las=2,xlab=ERR.LABEL,col=methods.colors[colnames(data2html$gene.err.sort)],horizontal=T)
  })
  par(mar=c(5,5,4,4))
  cat(r.a$html)
} 
%>
P<%=cur.plot%>
</TD></TR>
<TR><TD>
<%
#Error
diffs<-diag(data2html$sig.diff)
ndiffs.ttest<-diag(data2html$ttest.ndiff)
diffs.ttest<-diag(data2html$ttest.diff)
if ( sum(diffs) > 0 ) {
      cur.plot<-cur.plot+1
      r.a<-gen.plot2report(filename=paste(file.prefix,"_sigdiff_error.png",sep=""),
      width=max((length(diffs)*70),400),
      height=500,
      html=TRUE,
      ps=TRUE,
      to.plot=function() {
          barplot2(diffs[order(diffs)],log="y",ylab="Sig.diffs",las=2)
          barplot2(ndiffs.ttest[order(diffs)],las=2,add=T,col="darkgray")
          barplot2(diffs.ttest[order(diffs)],las=2,add=T,col="gray")
      })
      cat(r.a$html)
}
%>
P<%=cur.plot%>
</TD>
</tr>
<tr>
<TD>
<%

if ( nrow(data2html$p.cor)>0 ) {
   cur.plot<-cur.plot+1
   r.a<-gen.plot2report(filename=paste(file.prefix,"_sigdiff_posneg.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
                plot.b2b.barplot(diag(sig.neg),diag(sig.pos),groups=rownames(sig.neg))
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr><tr>
<TD>
<%

if ( is.Simulated ) {
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_errByexpr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
          caption="Genes sorted by average error (accross all pipelines)",
	  ps=TRUE,
          to.plot=function() {            
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
#            plot(smooth.spline(data2html$gene.err.sort[,1]),type="l",col=methods.colors[1],ylab=ERR.LABEL,xlab="Genes",log="y",xaxt="n",col=methods.colors[1])
             plot(smooth.spline(data2html$gene.err.sort[,1]),type="l",col=methods.colors[1],ylab=ERR.LABEL,xlab="Genes",ylim=c(0,200),xaxt="n")     
             for ( i in seq(2,length(colnames(data2html$gene.err.sort)))) {
                   lines(smooth.spline(data2html$gene.err.sort[,i]),col=methods.colors[colnames(data2html$gene.err.sort)[i]])
            }
            lines(smooth.spline(germ),type="l",col="black")
	    legend("topright", inset=c(-0.6,0), legend=append(colnames(data2html$gene.err.sort),"Mean"), lty=1, col=append(methods.colors[colnames(data2html$gene.err.sort)],"black"), title="", bty='n')
            axis(1,at=seq(1,nrow(data2html$gene.err.sort)+1,nrow(data2html$gene.err.sort)/5),labels=paste(seq(0,100,100/5),"%",sep=""))

      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>
<tr>
<TD>
<%

if ( is.Simulated ) {  
  cur.plot<-cur.plot+1
  quant.vals.u<-sort(unique(round(data2html$true.quant.mean.f,0)))
  quant.vals.r<-round(data2html$true.quant.mean.f,0)

  r.a<-gen.plot2report(filename=paste(file.prefix,"_expr_acc_err.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
              my.filt<-function(val,l1,l2) {
                sum(l2[l1>=val])
              }
            terror<-sum(data2html$gene.err.sort[,1])
            err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.err.sort[,1])
	    par(mar=c(5,5,4,10))
            plot(smooth.spline(err.by.expr/terror*100),type="l",log="x",xlab="Quantification (reads)",ylab="Accumulative error",ylim=c(0,100),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.err.sort)))) {
 	      terror<-sum(data2html$gene.err.sort[,i])
              err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.err.sort[,i])
              lines(smooth.spline(err.by.expr/terror*100),col=methods.colors[colnames(data2html$gene.err.sort)[i]])
            }
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.err.sort), lty=1, col=methods.colors[colnames(data2html$gene.err.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>
<tr>
<TD>
<%
if ( is.Simulated ) {  
  cur.plot<-cur.plot+1
  quant.vals.u<-sort(unique(round(data2html$true.quant.mean.f,0)))
  quant.vals.r<-round(data2html$true.quant.mean.f,0)
  r.a<-gen.plot2report(filename=paste(file.prefix,"_expr_err.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
	  caption="Average error VS expression",
          to.plot=function() {
              my.filt<-function(val,l1,l2) {
                mean(l2[l1==val])
              }
            err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.err.sort[,1])
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(smooth(err.by.expr),type="l",log="x",xlab="Quantification (reads)",ylab="avg. error",ylim=c(0,100),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.err.sort)))) {
              err.by.expr<-sapply(quant.vals.u,my.filt,quant.vals.r,data2html$gene.err.sort[,i])
              lines(smooth(err.by.expr),col=methods.colors[i])
            }
            # Add legend to top right, outside plot region
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.nerr.sort), lty=1, col=methods.colors[colnames(data2html$gene.nerr.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>

<tr>
<TD>
<%

if ( is.Simulated ) {
cur.plot<-cur.plot+1
sq<-names(sort(data2html$true.quant.mean.f))
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_errByexpr2.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          caption="Genes sorted by true quantification (counts)",
          to.plot=function() {
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="n");#reserve space for the legend
            plot(smooth.spline(data2html$gene.err.sort[sq,1]),type="l",ylab=ERR.LABEL,xlab="True quantification (counts)",xaxt="n",ylim=c(0,200),col=methods.colors[1])
            for ( i in seq(2,length(colnames(data2html$gene.err.sort)))) {
                   lines(smooth.spline(data2html$gene.err.sort[sq,i]),col=methods.colors[colnames(data2html$gene.err.sort)[i]])
            }
            lines(data2html$true.quant.mean.f[sq]/sum(data2html$true.quant.mean.f)*length(data2html$true.quant.mean.f)*100,col="gray")
	    legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.err.sort), lty=1, col=methods.colors[colnames(data2html$gene.err.sort)], title="", bty='n')
      })
cat(r.a$html)
}
%>
P<%=cur.plot%>
</td>
</tr>
<tr>
<TD>
<%

if ( is.Simulated ) {
cur.plot<-cur.plot+1
r.a<-gen.plot2report(filename=paste(file.prefix,"_gene_methods_errVsexpr.png",sep=""),
          width=500,
          height=500,
	  html=TRUE,
	  ps=TRUE,
          to.plot=function() {
	    par(mar=c(5,5,4,11),xpd=TRUE,bty="l");#reserve space for the legend
            plot(data2html$true.quant.mean.f,data2html$gene.err.sort[,1],log="xy",type="p",pch=19,xlab="True gene quantification (reads)",ylab=ERR.LABEL,col=methods.colors[1],cex=0.3)
            for ( i in seq(2,length(colnames(data2html$gene.err.sort))) ) { points(data2html$true.quant.mean.f,data2html$gene.err.sort[,i],col=methods.colors[i],pch=19,cex=0.3) }
          legend("topright", inset=c(-0.6,0), legend=colnames(data2html$gene.err.sort), pch=19, col=methods.colors[colnames(data2html$gene.err.sort)], title="", bty='n')
      })
cat(r.a$html)
#

}
%>
P<%=cur.plot%>
</TD></TR>


<TR><TD>
<!--
<%
library(R2HTML)
if ( nrow(data2html$p.cor)>0 ) {
  p.cor.html.out<-data2html$p.cor
  caption<-"Quantification correlation"
  HTML(p.cor.html.out, file = stdout(),
     Border = 0, innerBorder=0,
        classfirstline = "firstline", classfirstcolumn = "firstcolumn",
        classcellinside = "cellinside", append = FALSE,
        caption = caption, captionalign = "top", classcaption = "captiondataframe", classtable = "dataframe",
        digits=2, nsmall = 0, big.mark = " ", big.interval = 3, decimal.mark = ",",
        sortableDF=TRUE, row.names = FALSE)
} else { cat("") }
%>
-->
</TD></tr><tr><TD>
<!--
<%
if ( nrow(data2html$quant.diff)>0 ) {
  caption<-"Quantification difference"
  HTML(data2html$quant.diff, file = stdout(),
     Border = 0, innerBorder=0,
        classfirstline = "firstline", classfirstcolumn = "firstcolumn",
        classcellinside = "cellinside", append = FALSE,
        caption = caption, captionalign = "top", classcaption = "captiondataframe", classtable = "dataframe",
        digits=2, nsmall = 0, big.mark = " ", big.interval = 3, decimal.mark = ",",
        sortableDF=TRUE, row.names = FALSE)
} else { cat("") }
%>
-->
<%

%>
</TD></TR>
</TABLE>

<%# Comment -- wrap up everything in a div - end %>  
<!--</DIV>-->

<% brew::brew(get.path2template("page.footer")) %>

