<% brew::brew(get.path2template("page.header")) %>

<%
library(xtable)
%>
 
<%# Comment -- wrap up everything in a div %>
<DIV name='mapping_res'>
<H1>Mapping report for <%=opt$mapper%></H1>

<OL class="contents"> Contents
  <LI> <A href="#ms">Mapping statistics</A>
  <LI> <A href="#mp">Mapping profile</A>
</OL>


<H2 class="section">Mapping statistics<A NAME='ms'/></H2>

<H3 class="section"><%=irap.ctr("Figure")%></H3>
<%
file <- paste(out.prefix,"_aln",sep="")
n <- length(colnames(mapping.dfs$raw))
html <- gen.plot2report(filename=file,
                        dir=out.dir,
                        bg="white",
                        width=(480*round(0.5+n/12,0)),
                        height=480,
                        html=TRUE,
                        ps=TRUE,
                        caption="",
                        to.plot= function() {
                          mapping.plot(mapping.dfs$raw)
                        }
                        )

irap.output.html(html$html,"Alignment statistics per library: number of alignments, number of reads mapped, number of reads not mapped (unmapped), number of reads that map to a single locus (uniquely mapped), and number of reads spliced.")

%>

<H3 class="section"><%=irap.ctr("Figure")%></H3>
<%
file <- paste(out.prefix,"_diff",sep="")
n <- length(colnames(mapping.dfs$raw))
html <- gen.plot2report(filename=file,
                        dir=out.dir,
                        bg="white",
                        width=(480*round(0.5+n/12,0)),
                        height=480,
                        html=TRUE,
                        ps=TRUE,
                        caption="",
                        to.plot= function() {
                          mapping.qual(mapping.dfs$raw)
                        }
                        )

irap.output.html(html$html,"Number of alignments per library where a read aligns perfectly (no mismatches), 1 mismatch, 2 mismatches and more than 2 mismatches. The values are obtained from the BAM file(s).")
%>


<H3 class="section"><%=irap.ctr("Table")%></H3>
<%

xt <- xtable(mapping.dfs$raw,
             caption="",
             label="map_num_align_table")

html <- print.xtable(file="/dev/null",
                     xt,
                     type="html",
                     sanitize.text.function=function(x) {x},
                     include.colnames=TRUE,
                     include.rownames=TRUE,
                     html.table.attributes = "")

irap.output.html(html, "Alignment statistics per library: number of reads mapped, unmapped, uniquely mapped, multimaps and spliced; percentage of alignments where a read aligns pefectly (no mismatches), 1 mismatch, 2 mismatches, more than 2 mismatches, and spliced.")

tsv.download.files <- c(raw.stats.file)
names(tsv.download.files) <- "TSV"
abs.tsv.download.html <- html.download.bar(tsv.download.files)

%>
<DIV> <%=abs.tsv.download.html%></DIV>

  
<H2 class="section">Mapping profile<a name="mp"/></H2>
  
<H3 class="section"><%=irap.ctr("Figure")%></H3>
<%
rownames(mapping.feat)<-gsub("_"," ",rownames(mapping.feat))

file <- paste(out.prefix,"_aln_bio",sep="")
n <- length(colnames(mapping.feat))
mapping.feat2plot <- mapping.feat[,!colnames(mapping.feat) %in% c("Median","Mean","SD")]
html <- gen.plot2report(filename=file,
                        dir=out.dir,
                        bg="white",
                        width=(480*round(0.5+n/12,0)),
                        height=900,
                        html=TRUE,
                        ps=TRUE,
                        caption="",
                        to.plot= function() {
                          # sort rows based on mean values
                          feat.order <- names(sort(mapping.feat[,"Mean"],decreasing=T))
                          v <- mapping.feat2plot[feat.order,]
                          par(mar=c(9,5,4,10))
                          cols <- rainbow(nrow(mapping.feat2plot))
                          bp <- barplot(v,las=2,
                                        beside=F,col=cols,
                                        ylab="Number of alignments",
                                        horiz=F,cex.axis=0.8)

                          par( xpd=NA )
                          legend(ceiling(max(bp)),max(colSums(v)),legend=pprint.fieldname(rownames(v)),fill=cols,horiz=FALSE, cex=0.70,bty='n')                          
                        }
                        )

irap.output.html(html$html,"Number of alignments overlapping annotated features. Values obtained with bedtools (coverage). ")

%>


<H3 class="section"><%=irap.ctr("Table")%></H3>
<%
xt <- xtable(round(mapping.feat2plot,0),
             caption="",
             digits=0,
             label="map_feat_table")

html <- print.xtable(file="/dev/null",
                     xt,
                     type="html",
                     sanitize.text.function=function(x) {x},
                     include.colnames=TRUE,
                     include.rownames=TRUE,
                     html.table.attributes = "")

irap.output.html(html, "Number of alignments overlapping annotated features. Values obtained with bedtools (coverage).")

feat.download.files <- c(feat.stats.file)
names(feat.download.files) <- "TSV"
feat.tsv.download.html <- html.download.bar(feat.download.files)

%>
<DIV> <%=feat.tsv.download.html%></DIV>


<H3 class="section"><%=irap.ctr("Table")%></H3>
<%
mapping.feat.sum <- colSums(mapping.feat)
mapping.feat.per <- round(sweep(mapping.feat,MARGIN=2,100*1/mapping.feat.sum,`*`),2)
SD<-apply(mapping.feat.per[,-c(1,2,3)],1,FUN=sd)
mapping.feat.per[,"SD"]<-SD
#cbind(mapping.feat.per,SD)


xt <- xtable(mapping.feat.per,
             caption="",
             label="map_feat_table_per")

html <- print.xtable(file="/dev/null",
                     xt,
                     type="html",
                     sanitize.text.function=function(x) {x},
                     include.colnames=TRUE,
                     include.rownames=TRUE,
                     html.table.attributes = "")

irap.output.html(html, "Percentage of alignments overlapping annotated features. Values obtained with bedtools (coverage).")

%>

<H3 class="section"><%=irap.ctr("Figure")%></H3>
<%
rownames(mapping.feat.per)<-gsub("_"," ",rownames(mapping.feat.per))

file <- paste(out.prefix,"_paln_bio",sep="")
n <- length(colnames(mapping.feat.per))
mapping.feat2plot <- mapping.feat.per[,!colnames(mapping.feat.per) %in% c("Median","Mean","SD")]
html <- gen.plot2report(filename=file,
                        dir=out.dir,
                        bg="white",
                        width=(480*round(0.5+n/12,0)),
                        height=900,
                        html=TRUE,
                        ps=TRUE,
                        caption="",
                        to.plot= function() {
                          # sort rows based on mean values
                          feat.order <- names(sort(mapping.feat.per[,"Mean"],decreasing=T))
                          v <- mapping.feat2plot[feat.order,]
                          par(mar=c(9,5,4,10))
                          cols <- rainbow(nrow(mapping.feat2plot))
                          bp <- barplot(v,las=2,
                                        beside=F,col=cols,
                                        ylab="Percentage of alignments (%)",
                                        horiz=F,cex.axis=0.8)

                          par( xpd=NA )
                          legend(ceiling(max(bp)),max(colSums(v)),legend=pprint.fieldname(rownames(v)),fill=cols,horiz=FALSE, cex=0.70,bty='n')                          
                        }
                        )

irap.output.html(html$html,"Percentage of alignments per library overlapping annotated features. Values obtained with bedtools (coverage). ")

%>

  
<!-- number of reads mapping to exonic regions, intronic regions, other -->
<H3 class="section"><%=irap.ctr("Figure")%></H3>
<%

file <- paste(out.prefix,"_aln_genes",sep="")
mapping.genes2plot <- mapping.genes[,!colnames(mapping.genes) %in% c("Median","Mean","SD")]
n <- length(colnames(mapping.genes2plot))
html <- gen.plot2report(filename=file,
                        dir=out.dir,
                        bg="white",
                        width=(480*round(0.5+n/12,0)),
                        height=480,
                        html=TRUE,
                        ps=TRUE,
                        caption="",
                        to.plot= function() {
                          cols <- rainbow(nrow(mapping.genes2plot))
                          par(mar=c(10,5,5,4))                          
                          bp <- barplot(mapping.genes2plot,las=2,col=cols,)
                          par( xpd=NA )                          
                          legend("top",inset=c(0,-0.08),legend=rownames(mapping.genes2plot),fill=cols,horiz=TRUE, cex=0.80,bty='n')                            
                        }
                        )

irap.output.html(html$html,"Number of alignments overlapping introns and exons. Values obtained with bedtools (intersect).")

%>

<H3 class="section"><%=irap.ctr("Table")%></H3>
<%
save.image("debug.Rdata")
head(mapping.genes)
xt <- xtable(round(mapping.genes[,!colnames(mapping.genes) %in% c("Median","Mean","SD")],0),
             caption="",
             digits=0,
             label="map_genes_table")

html <- print.xtable(file="/dev/null",
                     xt,
                     type="html",
                     sanitize.text.function=function(x) {x},
                     include.colnames=TRUE,
                     include.rownames=TRUE,
                     html.table.attributes = "")

irap.output.html(html, "Number of alignments overlapping  exons and introns. Values obtained with bedtools (intersect).")

gene.download.files <- gene.stats.file
names(gene.download.files) <- "TSV"
gene.tsv.download.html <- html.download.bar(gene.download.files)

%>
<DIV> <%=gene.tsv.download.html%></DIV>

<H3 class="section"><%=irap.ctr("Table")%></H3>  
<%
# percentages
tot <- colSums(mapping.genes)
p.mapping.genes <- round(sweep(mapping.genes,MARGIN=2,100*1/tot,`*`),2)
# recompute sd
SD<-apply(p.mapping.genes[,-c(1,2,3)],1,FUN=sd)
p.mapping.genes[,"SD"] <- SD
#p.mapping.genes<-cbind(p.mapping.genes,SD)
xt <- xtable(p.mapping.genes,
             caption="",
             label="map_genes_table.p")

html <- print.xtable(file="/dev/null",
                     xt,
                     type="html",
                     sanitize.text.function=function(x) {x},
                     include.colnames=TRUE,
                     include.rownames=TRUE,
                     html.table.attributes = "")

irap.output.html(html, "Percentage of alignments overlapping exons and introns. Values obtained with bedtools.")

%>





</DIV>

<% brew::brew(get.path2template("page.footer")) %>

